arroyo/processing/strategies/router.py

```diff
@@ -0,0 +1,373 @@
+import logging
+from collections import deque
+from functools import partial
+from typing import Deque, Mapping, MutableMapping, Optional, Protocol, Set, Union
+
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Commit, FilteredPayload, Message, Partition, TStrategyPayload
+
+logger = logging.getLogger(__name__)
+
+
+class PartitionWatermark:
+    """
+    Keeps track of the highest committable offset when messages are processed
+    by a discrete set of independent strategies (routes here) which can
+    issue commit according to different strategies, thus out of order.
+
+    Example:
+
+    ```
+    Offsets produced:
+    Route 1:  ---- 10 --- 15 --- 20 --------------------
+    Route 2:  ------------------------- 25 --- 30 --- 35
+    ```
+
+    Route1 offsets are processed by a different strategy independently from
+    Route2 offsets. This makes it is totally possible that the strategy that
+    processes route 2 would commit offset 25, 30, 35 before route 1 commits 20.
+
+    This class keeps track that the highest committable offset is not going
+    to be 35 until route one issued the commit for offset 20.
+
+    This class has to keep track of all offsets above the highest committable
+    at a given point in time in order as, depending on the order of the
+    received commits, the watermark can land on any of the offsets. We cannot
+    keep only the highest offset per route.
+    """
+
+    def __init__(self, routes: Set[str]):
+        self.__committed: Mapping[str, Deque[int]] = {
+            route: deque() for route in routes
+        }
+        self.__uncommitted: Mapping[str, Deque[int]] = {
+            route: deque() for route in routes
+        }
+
+        self.__highest_committed: Optional[int] = None
+
+    def add_message(self, route: str, offset: int) -> None:
+        """
+        Adds one uncommitted offset to one route.
+
+        An uncommitted offset is added when the message is consumed, it is
+        being processed but before the offset is committed.
+        """
+        self.__uncommitted[route].append(offset)
```

Comment:
A small idea:

nit: Instead of growing/shrinking, pre-allocate a buffer and use it as storage with 2 pointers, and depending on the usage, increase it (like an arena allocator)